### PR TITLE
Add extra WASM heap pages when precompiling the runtime blob

### PR DESCRIPTION
**Pull request by koute**
polkadot companion: https://github.com/paritytech/polkadot/pull/5197

Fixes https://github.com/paritytech/polkadot/issues/5190

### CODE REVIEW

File: client/executor/wasmtime/src/tests.rs [hunk 1]
```diff
@@ -168,24 +178,36 @@ fn test_stack_depth_reaching() {
...
 		wat: String,
+		precompile_runtime: bool,
 	) -> Result<(), Box<dyn std::error::Error>> {
 		let runtime = {
 			let mut builder = RuntimeBuilder::new_on_demand();
 			builder.use_wat(wat);
 			builder.max_memory_size(max_memory_size);
+			builder.precompile_runtime(precompile_runtime);
 			builder.build()
 		};
```
Comment (wigy-opensource-developer) [inline client/executor/wasmtime/src/tests.rs:212]: I know it was not you who introduced this builder, but the best practice for the [builder pattern in Rust](https://rust-unofficial.github.io/patterns/patterns/creational/builder.html) is that each method takes `self` and returns it back. That allows one to write:

```rust
let runtime = RuntimeBuilder::new_on_demand()
	.use_wat(wat)
	.max_memory_size(max_memory_size)
	.precompile_runtime(precompile_runtime)
	.build();
```

**Review by wigy-opensource-developer — APPROVED** [review]
Besides the rustdoc, LGTM

**Comment by wigy-opensource-developer** [comment]
I was too fast. You fixed the rustdoc while I was reviewing :wink:

File: client/executor/wasmtime/src/tests.rs (hunk 1 above)
Comment (wigy-opensource-developer) [reply client/executor/wasmtime/src/tests.rs:212]: Okay, you can use fluent API like I wrote with `&mut self` arguments, too.

**Review by KiChjang — APPROVED** [review]

**Review by bkchr — CHANGES_REQUESTED** [review]
I think this is incorrect. Maybe it was already incorrect before, but the heap pages should be the maximum. So, if polkadot it setting these for parachains it is 1gb and nothing in the wasm file should allow to exceed this value. Currently the wasm file could specify whatever initial value it wants and it would get this plus the heap pages we are requesting by polkadot?

For normal runtimes this isn't such a big problem, as they are not provided by the user.

**Comment by koute** [comment]
> I think this is incorrect. Maybe it was already incorrect before, but the heap pages should be the maximum.

This is indeed how it worked originally, I just changed the name of the field to actually say what it does.

We can consider changing it, and I agree with you that changing this would make sense (which is [what I wrote in this comment](https://github.com/paritytech/substrate/pull/10480#discussion_r803408345)), but I think that'd be better for another PR as this just restores the original behavior from before the https://github.com/paritytech/substrate/pull/10480 landed.

> So, if polkadot it setting these for parachains it is 1gb and nothing in the wasm file should allow to exceed this value. Currently the wasm file could specify whatever initial value it wants and it would get this plus the heap pages we are requesting by polkadot?
> 
> For normal runtimes this isn't such a big problem, as they are not provided by the user.

Yes, the blob can request whatever value it wants, and it will get that amount of heap pages plus the `extra_heap_pages`. But there's still an absolute limit set by the `Config::max_memory_size`. If it goes over that then the instantiation **will** fail, so in practice this is - while weird - not a problem. (For PVFs the limit is set [here](https://github.com/paritytech/polkadot/pull/5197/files#diff-f4ffe6a7003b388a382a12c2ac95731f8862f4e3cd96be2898f271c131ee528aR44), so with the current limit a PVF can get at most 2080 pages allocated, which is ~130MB of memory.)

**Review by bkchr — APPROVED** [review]
Okay, sounds good.

**Pull request merged**